src/Raven.Client/Http/RequestExecutor.cs

```diff
@@ -639,220 +641,226 @@ protected virtual bool TryLoadFromCache(JsonOperationContext context)
+                                        if (shouldRetry == false)
+                                            throw timeoutException;
 
-                                response = await preferredTask.ConfigureAwait(false);
+                                        sp.Stop();
 
-                                if (TryGetServerVersion(response, out var serverVersion))
-                                    LastServerVersion = serverVersion;
+                                        if (sessionInfo != null)
+                                            sessionInfo.AsyncCommandRunning = false;
 
-                                if (sessionInfo?.LastClusterTransactionIndex != null)
-                                {
-                                    // if we reach here it means that sometime a cluster transaction has occurred against this database.
-                                    // Since the current executed command can be dependent on that, we have to wait for the cluster transaction.
-                                    // But we can't do that if the server is an old one.
-                                    if (serverVersion == null || string.Compare(serverVersion, "4.1", StringComparison.Ordinal) < 0)
-                                        throw new ClientVersionMismatchException(
-                                            $"The server on {chosenNode.Url} has an old version and can't perform the command '{command.GetType()}', " +
-                                            "since this command dependent on a cluster transaction which this node doesn't support");
+                                        if (await HandleServerDown(url, chosenNode, nodeIndex, context, command, request, response, timeoutException, sessionInfo, token)
+                                                .ConfigureAwait(false) == false)
+                                            ThrowFailedToContactAllNodes(command, request);
+
+                                        return;
+                                    }
+
+                                    throw;
                                 }
                             }
-                            catch (OperationCanceledException e)
+                            else
                             {
-                                if (cts.IsCancellationRequested && token.IsCancellationRequested == false) // only when we timed out
-                                {
-                                    var timeoutException = new TimeoutException($"The request for {request.RequestUri} failed with timeout after {timeout}", e);
-                                    if (shouldRetry == false)
-                                        throw timeoutException;
-
-                                    sp.Stop();
+                                cts.CancelAfter(DefaultHttpClientTimeout);
```

Comment:
What does this gives us that the usual timeout doesn't?

Also, note that this doesn't handle timeouts on read, which is the root cause of our problem